### PR TITLE
evaluator: change interface to datum in evaluator.

### PR DESCRIPTION
**Pull request by zxylvlp**
Change the remaining functions.

### CODE REVIEW

**Comment by zxylvlp** [comment]
@coocood @shenli @zimulala

File: evaluator/builtin_control.go [hunk 1]
```diff
@@ -19,20 +19,21 @@ import (
 )
 
 // See https://dev.mysql.com/doc/refman/5.7/en/control-flow-functions.html#function_if
-func builtinIf(args []interface{}, _ context.Context) (interface{}, error) {
+func builtinIf(args []types.Datum, _ context.Context) (types.Datum, error) {
```
Comment (zimulala) [inline evaluator/builtin_control.go]: need `(types.Datum, error)` -> `(d types.Datum, err error)` like other builtin functions ?

File: util/types/datum.go [hunk 2]
```diff
@@ -1078,6 +1088,49 @@ func (d *Datum) ToBool() (int64, error) {
 	return 1, nil
 }
 
+// ConvertDatumToDecimal converts interface to decimal.
```
Comment (coocood) [inline util/types/datum.go]: converts Datum

File: evaluator/evaluator_binop.go [hunk 3]
```diff
@@ -281,284 +281,369 @@ func (e *Evaluator) handleArithmeticOp(o *ast.BinaryOperationExpr) bool {
...
-	case []byte:
+		du := a.GetMysqlDuration()
+		de := du.ToNumber()
+		if du.Fsp == 0 {
+			d.SetInt64(de.IntPart())
+			return d, nil
+		}
+		d.SetMysqlDecimal(de)
+		return d, nil
+	case types.KindBytes:
```
Comment (coocood) [inline evaluator/evaluator_binop.go]: KindBytes and KindString can use the same code.

File: util/types/datum.go [hunk 4]
```diff
@@ -1078,6 +1088,49 @@ func (d *Datum) ToBool() (int64, error) {
...
+		return mysql.NewDecimalFromUint(uint64(d.GetMysqlSet().Value), 0), nil
+	default:
+		return mysql.Decimal{}, fmt.Errorf("can't convert %v to decimal", d.GetValue())
+	}
+}
+
+// ToDecimal converts to a decimal.
+func (d *Datum) ToDecimal() (mysql.Decimal, error) {
+	switch d.Kind() {
+	case KindBytes:
```
Comment (coocood) [inline util/types/datum.go]: This case can be removed, and add KindBytes after KindString case above.

File: evaluator/evaluator_binop.go [hunk 5]
```diff
@@ -281,284 +281,361 @@ func (e *Evaluator) handleArithmeticOp(o *ast.BinaryOperationExpr) bool {
...
-
-	return types.InvOp2(a, b, opcode.Minus)
+func computeMinus(a, b types.Datum) (d types.Datum, err error) {
+	switch a.Kind() {
+	case types.KindInt64:
+		switch b.Kind() {
+		case types.KindInt64:
+			r, err1 := types.SubInt64(a.GetInt64(), b.GetInt64())
+			d.SetInt64(r)
+			return d, err1
```
Comment (shenli) [inline evaluator/evaluator_binop.go]: If err1 != nil, we should use errors.Trace(err1)

**Comment by shenli** [comment]
LGTM
@coocood @zimulala

File: evaluator/builtin_control.go [hunk 6]
```diff
@@ -14,6 +14,7 @@
 package evaluator
 
 import (
+	"github.com/pingcap/tidb/Godeps/_workspace/src/github.com/juju/errors"
```
Comment (coocood) [inline evaluator/builtin_control.go]: Wrong import

File: evaluator/evaluator.go [hunk 7]
```diff
@@ -535,16 +551,16 @@ func (e *Evaluator) unaryOperation(u *ast.UnaryOperationExpr) bool {
...
+			u.SetMysqlDecimal(mysql.ZeroDecimal.Sub(aDatum.GetMysqlTime().ToNumber()))
 		case types.KindString:
 			f, err := types.StrToFloat(aDatum.GetString())
 			e.err = errors.Trace(err)
 			u.SetFloat64(-f)
 		case types.KindMysqlDecimal:
 			f, _ := aDatum.GetMysqlDecimal().Float64()
-			u.SetValue(mysql.NewDecimalFromFloat(-f))
+			u.SetMysqlDecimal(mysql.NewDecimalFromFloat(-f))
 		case types.KindBytes:
```
Comment (coocood) [inline evaluator/evaluator.go]: Merge this case with KindString above.

**Comment by coocood** [comment]
LGTM

**Pull request merged**